### ISSUE TITLE
"ZeroMQ version not supported!"

### ISSUE THREAD

**Issue by adzenith** [zeromq/cppzmq#152]
I have been trying to figure this out for some time now, and I can't for the life of me.

Steps to reproduce: the ones in the README, to the letter.

tar files:
`https://github.com/zeromq/libzmq/archive/v4.2.2.tar.gz`
`https://github.com/zeromq/cppzmq/archive/v4.2.2.tar.gz`

libzmq builds and install fine. No matter what I do, cppzmq then doesn't configure:

```
-- The C compiler identification is GNU 5.4.0
-- The CXX compiler identification is GNU 5.4.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
CMake Error at CMakeLists.txt:16 (message):
  ZeroMQ version not supported!
```

Any tips?

**Comment by bluca** [comment]
Strange, are you on Linux?

**Comment by nikoargo** [comment]
Yeah, Ubuntu 16.04.

**Comment by bluca** [comment]
I would recommend to use the packages then:

https://software.opensuse.org//download.html?project=network%3Amessaging%3Azeromq%3Agit-stable&package=libzmq3-dev

As it might be a problem of non-standard install path

**Comment by adzenith** [comment]
Yeah, I'm trying to install it somewhere nonstandard, but even without that it's not working (and it was successfully done [here](https://github.com/zeromq/cppzmq/pull/128#issuecomment-310892175)).

Here's a minimal example as a Dockerfile. I run it with `docker build -f Dockerfile .`.

```
FROM ubuntu:16.04

ENV VERSION=4.2.2

# Zeromq Dependencies
RUN apt-get update && apt-get install -y \
    cmake \
    g++ \
    wget

# Pull source and extract
RUN mkdir -p /tmp \
    && cd /tmp \
    && wget -O libzmq-$VERSION.tar.gz https://github.com/zeromq/libzmq/archive/v$VERSION.tar.gz \
    && tar -xvf libzmq-$VERSION.tar.gz \
    && wget -O cppzmq-$VERSION.tar.gz https://github.com/zeromq/cppzmq/archive/v$VERSION.tar.gz \
    && tar -xvf cppzmq-$VERSION.tar.gz

# Build
RUN cd /tmp/libzmq-$VERSION \
    && mkdir build \
    && cd build \
    && cmake .. -DCMAKE_INSTALL_PREFIX="/usr/local/zeromq-$VERSION" \
    && make -j 10 \
    && make install

RUN cd /tmp/cppzmq-$VERSION \
    && mkdir build \
    && cd build \
    && cmake .. -DCMAKE_INSTALL_PREFIX="/usr/local/zeromq-$VERSION" \
    && make -j 10 \
    && make install
```

**Comment by bluca** [comment]
It's probably missing some CMake variable to set the default search path. Have you tried with the packages?

**Comment by adzenith** [comment]
The problem isn't that it's not finding the version, it's that it says `ZeroMQ version not supported!` - as far as I can telling from adding `message`s to the CMake, it's finding it just fine.

The reason we're not using packages right now is we're trying to install ZMQ at a custom prefix. I don't think that's possible with the packages.

**Comment by bluca** [comment]
are you sure you don't have something like libzmq1 installed as a package?

**Comment by adzenith** [comment]
Yeah - that Dockerfile is based on the `ubuntu:16.04` image, which is a super minimal image with almost nothing in it. I just printed out all the packages in it and there's no zmq there (`apt list --installed | grep zmq`).

Is there any more debugging info that I can get? I was trying to figure out how to get CMake to tell me what package it had found, but CMake is pretty obtuse.

**Comment by bluca** [comment]
It is pretty obtuse, and I'm afraid I don't know what else you could try, sorry, maybe someone else can

As a workaround, consider that this repository doesn't really install much - it's just 2 headers, so you can just copy them in the include/ directory that you want, if that's all you need

**Comment by adzenith** [comment]
Oh ok, great! I'll just copy them by hand.

The CMakeLists.txt says things like `add_library`, etc., which make me think it was doing more than just copying headers.

Thanks!

**Comment by andyshi2008** [comment]
I have the same problem on Mac, libzmq4.2.2. and cppzmq4.2.2 . using @bluca method workaroud.

**Comment by singularTnT** [comment]
When I configure the cppzmq, error occurs like this:

sudo cmake .. -DCMAKE_INSTALL_PREFIX=/tmp/test
CMake Error at /home/lwp/anaconda3/share/cmake/ZeroMQ/ZeroMQConfig.cmake:44 (include):
  include could not find load file:

    /home/lwp/anaconda3/share/cmake/ZeroMQ/ZeroMQTargets.cmake
Call Stack (most recent call first):
  CMakeLists.txt:4 (find_package)


CMake Error at /home/lwp/anaconda3/share/cmake/ZeroMQ/ZeroMQConfig.cmake:46 (get_target_property):
  get_target_property() called with non-existent target "libzmq".
Call Stack (most recent call first):
  CMakeLists.txt:4 (find_package)


CMake Error at /home/lwp/anaconda3/share/cmake/ZeroMQ/ZeroMQConfig.cmake:47 (get_target_property):
  get_target_property() called with non-existent target "libzmq".
Call Stack (most recent call first):
  CMakeLists.txt:4 (find_package)


CMake Error at /home/lwp/anaconda3/share/cmake/ZeroMQ/ZeroMQConfig.cmake:48 (get_target_property):
  get_target_property() called with non-existent target "libzmq-static".
Call Stack (most recent call first):
  CMakeLists.txt:4 (find_package)


CMake Error at CMakeLists.txt:16 (message):
  ZeroMQ version not supported!


-- Configuring incomplete, errors occurred!
See also "/home/lwp/Software/cppzmq-master/build/CMakeFiles/CMakeOutput.log".

Any ideas? Please.

**Comment by siretty** [comment]
Hey, I just ran into the same issue. It seems as if (in installations from the git version - in contrast to the POSIX tar file where cmake builds are currently broken) the targets libzmq and libzmq-static are not defined. You can fix the ZeroMQConfig.cmake file yourself (if you can live with that rather ugly hotfix) or open an issue for ZeroMQ.

For me applying the following patch did the trick (generous application of -pN is recommended):
````patch
diff --git a/deps/root/share/cmake/ZeroMQ/ZeroMQConfig.cmake b/deps/root/share/cmake/ZeroMQ/ZeroMQConfig.cmake                                                                                                
index d325447..5f751d9 100644
--- a/deps/root/share/cmake/ZeroMQ/ZeroMQConfig.cmake
+++ b/deps/root/share/cmake/ZeroMQ/ZeroMQConfig.cmake
@@ -37,3 +37,13 @@ set_and_check(${PN}_INCLUDE_DIR "${PACKAGE_PREFIX_DIR}/include")
 set_and_check(${PN}_LIBRARY "${PACKAGE_PREFIX_DIR}/lib64/libzmq.so")
 set_and_check(${PN}_STATIC_LIBRARY "${PACKAGE_PREFIX_DIR}/lib64/libzmq.a")
 check_required_components(${PN})
+
+if(NOT TARGET libzmq) # in case find_package is called multiple times
+  add_library(libzmq SHARED IMPORTED)
+  set_target_properties(libzmq PROPERTIES IMPORTED_LOCATION ${${PN}_LIBRARY})
+endif(NOT TARGET libzmq)
+
+if(NOT TARGET libzmq-static) # in case find_package is called multiple times
+  add_library(libzmq-static STATIC IMPORTED)
+  set_target_properties(libzmq-static PROPERTIES IMPORTED_LOCATION ${${PN}_STATIC_LIBRARY})
+endif(NOT TARGET libzmq-static)
````

**Comment by sigiesec** [comment]
This was fixed by #191, thanks for the suggestion @siretty !
